crates/vm/src/arch/extensions.rs

```diff
@@ -966,15 +977,33 @@ impl<F: PrimeField32, E, P> VmChipComplex<F, E, P> {
     pub(crate) fn generate_proof_input<SC: StarkGenericConfig>(
         mut self,
         cached_program: Option<CommittedTraceData<SC>>,
+        trace_height_constraints: &[LinearConstraint],
         #[cfg(feature = "bench-metrics")] metrics: &mut VmMetrics,
-    ) -> ProofInput<SC>
+    ) -> Result<ProofInput<SC>, GenerationError>
     where
         Domain<SC>: PolynomialSpace<Val = F>,
         E: Chip<SC>,
         P: AnyEnum + Chip<SC>,
     {
         // System: Finalize memory.
         self.finalize_memory();
+
+        let trace_heights = self.current_trace_heights();
+        if any(&trace_heights, |h| *h > self.max_trace_height) {
+            return Err(GenerationError::TraceHeightsLimitExceeded);
+        }
+        if trace_height_constraints.is_empty() {
+            tracing::warn!("generating proof input without trace height constraints");
+        }
+        for constraint in trace_height_constraints {
+            let value = zip_eq(&constraint.coefficients, &trace_heights)
+                .map(|(c, h)| *c as u64 * *h as u64)
+                .sum::<u64>();
+            if value >= constraint.threshold as u64 {
+                return Err(GenerationError::TraceHeightsLimitExceeded);
```

Comment:
for debugging, either error with the constraint or constraint index? or log something